js/util/data-fetcher.js

```diff
@@ -0,0 +1,105 @@
+/**
+ * Copyright (C) 2015-2018 Regents of the University of California.
+ * @author: Chavoosh Ghasemi <chghasemi@cs.arizona.edu>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ * A copy of the GNU Lesser General Public License is in the file COPYING.
+ */
+
+/** @ignore */
+var Interest = require('../interest.js').Interest; /** @ignore */
+var NdnCommon = require('./ndn-common.js').NdnCommon;
+var LOG = require('../log.js').Log.LOG;
+
+/**
+ * DataFetcher is a utility class to resolve a given segment.
+ *
+ * This is a public constructor to create a new DataFetcher object.
+ * @param {Pipeline} pipe This is a pipeline that is in charge of retrieving
```

Comment:
Right, I forgot to remove them.

Reply:
To be clear: you need to remove `pipe`, not the callback functions.

> all pipelines MUST implement onData, onNack, and onTimeout methods

This feels like Java. Early NDN-JS (around 2013) had upcalls like this everywhere. We have since moved away from upcalls.
In JavaScript, the API should feel like JavaScript. Therefore, callbacks or emitters, no upcalls or interfaces.

Reply:
The problem with callback is `this` pointer. When I do callback in `DataFetcher` class (assume `onData`), `this` in `onData` class in `PipelineFixed` does not point to the right object and the program fails. That is why I use upcall.

Reply:
It's time to learn a bit of JavaScript fundamentals: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/bind

I'm not laughing at you. I have been there. My first JavaScript app doesn't even have objects. It has 500 global variables.

Reply:
Thank you. I fixed it. Let me know if still there is a problem.